### PR TITLE
ARUHA-164: reject fields if they are not in the schema

### PR DESCRIPTION
**Pull request by v1ctor**

### CODE REVIEW

File: src/test/java/de/zalando/aruha/nakadi/utils/TestUtils.java [hunk 1]
```diff
@@ -100,9 +102,16 @@ public static String resourceAsString(final String resourceName, final Class cla
     }
 
     public static EventType buildEventType(final String name, final JSONObject schema) {
+        return buildEventType(name, schema, Collections.emptyList());
+    }
+
+    public static EventType buildEventType(final String name, final JSONObject schema, List<String>
```
Comment (v-stepanov) [inline src/test/java/de/zalando/aruha/nakadi/utils/TestUtils.java]: final

**Comment by v-stepanov** [comment]
I'm not sure if we can merge it right now. I think some of users will have problems with that. If I remember correctly eventlog guys were using metadata.eid as a partitioning key. (which is not part of schema of course)

File: src/test/java/de/zalando/aruha/nakadi/controller/EventTypeControllerTest.java [hunk 2]
```diff
@@ -264,9 +266,23 @@ public void whenDeleteEventTypeAndTopicDeletionExceptionThen503() throws Excepti
     }
 
     @Test
+    public void whenCreateEventTypeWithWrongPartitionKeyFieldsThen422() throws Exception {
+
+        final EventType eventType = buildEventType(randomValidEventTypeName(), new JSONObject("{ \"price\": 1000 }"),
+                Collections.singletonList("blabla"));
+
+        Mockito.doReturn(eventType).when(eventTypeRepository).findByName(eventType.getName());
+        Mockito.doThrow(new TopicDeletionException("dummy message", null)).when(topicRepository).deleteTopic(
```
Comment (rcillo) [inline src/test/java/de/zalando/aruha/nakadi/controller/EventTypeControllerTest.java]: I could not understand why is this mock `doThrow` needed. Are you sure it's needed?

File: src/main/java/de/zalando/aruha/nakadi/controller/EventTypeController.java [hunk 3]
```diff
@@ -192,6 +193,13 @@ private void validateSchema(final EventType eventType) throws InvalidEventTypeEx
                 throw new InvalidEventTypeException("\"metadata\" property is reserved");
             }
 
+            List<String> absentFields = eventType.getPartitionKeyFields().stream()
```
Comment (rcillo) [inline src/main/java/de/zalando/aruha/nakadi/controller/EventTypeController.java]: Is the update flow also safe? Or just creation flow?

**Comment by rcillo** [comment]
@v-stepanov regarding your comment, this should no longer work for data change events, for example, since the partition key lookup is done scoped by "data" field.

So I think it's fine to release this feature.

File: src/main/java/de/zalando/aruha/nakadi/controller/EventTypeController.java (hunk 3 above)
Comment (v1ctor) [inline src/main/java/de/zalando/aruha/nakadi/controller/EventTypeController.java]: We can't change schema or partition keys in update. So it's working only for update.

File: src/test/java/de/zalando/aruha/nakadi/controller/EventTypeControllerTest.java (hunk 2 above)
Comment (v1ctor) [inline src/test/java/de/zalando/aruha/nakadi/controller/EventTypeControllerTest.java]: You right, I've copied it form other test.

File: src/main/java/de/zalando/aruha/nakadi/controller/EventTypeController.java (hunk 3 above)
Comment (rcillo) [inline src/main/java/de/zalando/aruha/nakadi/controller/EventTypeController.java]: Are you sure it's not possible to change partition keys in updates?

Comment (v1ctor) [inline src/main/java/de/zalando/aruha/nakadi/controller/EventTypeController.java]: You are right. There is a possibility to change partition keys. I will fix it.

**Comment by v-stepanov** [comment]
@rcillo but this still can be a problem for business events
and I think all eventlog events are business events
@v1ctor for me it would be more comfortable if we have a feature toggle for this.

**Comment by rcillo** [comment]
@v-stepanov I see. So maybe this validation should be done based on the effective schema?

**Comment by v1ctor** [comment]
@v-stepanov I've added feature toggle for that

**Comment by rcillo** [comment]
:shipit:

**Pull request merged**